# 제 6회 기출 문제

## 작업형 제 1 유형

### ✏️ 출동시간과 도착시간 차이가 평균적으로 가장 오래 걸린 소방서의 시간을 분으로 변환하여 출력하시오.

In [1]:
import pandas as pd
df = pd.read_csv('./data/data6-1-1.csv')
print(df.head())

    소방서                 출동시간                 도착시간
0  소방서1  2023-09-16 06:12:29  2023-09-16 07:51:55
1  소방서5  2023-09-19 06:11:41  2023-09-19 08:08:36
2  소방서1  2023-09-14 06:12:00  2023-09-14 07:24:21
3  소방서6  2023-09-20 06:12:24  2023-09-20 06:30:45
4  소방서9  2023-09-15 06:11:50  2023-09-15 06:44:00


In [2]:
# 1. datetime 형태로 변경
## 도착시간과 출동시간을 object 에서 datetime 자료형으로 변경
df['출동시간'] = pd.to_datetime(df['출동시간'])
df['도착시간'] = pd.to_datetime(df['도착시간'])

# 2. 출동시간과 도착시간 차이를 분으로 계산
## dt.total_seconds() 함수는 전체 시간을 초 단위로 변환 -> 60으로 나누면 분
df['시간차이(분)'] = (df['도착시간'] - df['출동시간']).dt.total_seconds() / 60
print(df.head(), '\n')

# 3. 소방서별 평균 시간 차이 계산
## 소방서별 시간 차이의 평균 구하기
## groupby('소방서') 에서는 소괄호를 사용해 그룹핑 기준 컬럼 선택
## ['시간차이(분)'].mean() 활용해 시간 차이(분) 컬럼의 평균만 구하기
avg_diff = df.groupby('소방서')['시간차이(분)'].mean()
print(avg_diff, '\n')

# 4. 평균 차이가 가장 큰 소방서의 시간을 찾고 출력
## idxmax() 는 가장 큰 값의 인덱스명(값)을 받환받음
## '소방서9' 값을 갖고 있는 인덱스의 값 출력
## 출력 결과를 round() 로 반올림해 정수 출력
idx = avg_diff.idxmax()
result = avg_diff[idx]
print(result)
print(round(result))

    소방서                출동시간                도착시간     시간차이(분)
0  소방서1 2023-09-16 06:12:29 2023-09-16 07:51:55   99.433333
1  소방서5 2023-09-19 06:11:41 2023-09-19 08:08:36  116.916667
2  소방서1 2023-09-14 06:12:00 2023-09-14 07:24:21   72.350000
3  소방서6 2023-09-20 06:12:24 2023-09-20 06:30:45   18.350000
4  소방서9 2023-09-15 06:11:50 2023-09-15 06:44:00   32.166667 

소방서
소방서1     54.952381
소방서10    42.365000
소방서2     60.228571
소방서3     43.490741
소방서4     49.136667
소방서5     61.250000
소방서6     41.764583
소방서7     56.670370
소방서8     42.693750
소방서9     80.658889
Name: 시간차이(분), dtype: float64 

80.6588888888889
81


---

#### 💡 분의 소수점 반올림
- result 결과값은 80.6588888888889

    - 80이 분, 소수점 이하는 초로 이해하는 경우가 많지만 아님
    
    - 80도 분, 소수점 이하도 분
    
        - '분'에 round() 함수로 반올림하는 것이 정상

In [3]:
min = 5.5       # 5도 '분', 0.5도 '분
print(int(min), '분')
print((min - int(min)) * 60, '초')  # 0.5 를 초로 재계산하면 30초

5 분
30.0 초


---

### ✏️ 학교에서 교사 한 명당 맡은 학생 수가 가장 많은 학교를 찾고, 그 학교의 전체 교사 수를 구하시오.

In [4]:
import pandas as pd
df = pd.read_csv('./data/data6-1-2.csv')
print(df.head())

   학교명  교사수  1학년  2학년  3학년  4학년  5학년  6학년
0  학교1   20  118   72  154  118   77   59
1  학교2   31  187  193  135  137   63   90
2  학교3   23  136  193  135  112   91   50
3  학교4   32   72  146   67   51  130  164
4  학교5   22   61   73   86   82   55  124


In [5]:
# [풀이1]
# 1. 총 학생 수 계산
## 1학년부터 6학년까지 더한 값을 총 학생 수로 계산
df['총학생수'] = df.iloc[:, 2:].sum(axis=1)
print(df, '\n')

# 2. 교사 한 명단 맡은 학생 수
## 총 학생 수 / 교사 수로 교사 한 명당 맡은 학생의 수 계산
df['학생/교사'] = df['총학생수'] / df['교사수']
print(df.head(), '\n')

# 3. 학생/교사 컬럼을 내림차순으로 정렬
## '학생/교사' 컬럼을 기준으로 내림차순으로 정렬하면 '학생/교사' 컬럼에서 가장 큰 값이 첫 번째 행
df = df.sort_values('학생/교사', ascending=False)
print(df.head(), '\n')

# 4. 최상단 행의 교사 수 값 출력
## 정렬을 하고 나면 인덱스명(값)이 뒤섞이게 되므로 iloc 사용
## 첫번째 행의 교사 수 값을 출력
print(df.iloc[0, 1])

      학교명  교사수  1학년  2학년  3학년  4학년  5학년  6학년  총학생수
0     학교1   20  118   72  154  118   77   59   598
1     학교2   31  187  193  135  137   63   90   805
2     학교3   23  136  193  135  112   91   50   717
3     학교4   32   72  146   67   51  130  164   630
4     학교5   22   61   73   86   82   55  124   481
..    ...  ...  ...  ...  ...  ...  ...  ...   ...
95   학교96   38   75  166  113  191  172   57   774
96   학교97   30  125  169  145  122  187  194   942
97   학교98   37  136  100   93   60  145   63   597
98   학교99   37   74  175  145  157   95   76   722
99  학교100   40   68  109  196  148   84  179   784

[100 rows x 9 columns] 

   학교명  교사수  1학년  2학년  3학년  4학년  5학년  6학년  총학생수      학생/교사
0  학교1   20  118   72  154  118   77   59   598  29.900000
1  학교2   31  187  193  135  137   63   90   805  25.967742
2  학교3   23  136  193  135  112   91   50   717  31.173913
3  학교4   32   72  146   67   51  130  164   630  19.687500
4  학교5   22   61   73   86   82   55  124   481  21.863636 

     학

In [6]:
import pandas as pd
df = pd.read_csv('./data/data6-1-2.csv')

# [풀이2]
# 1. 총 학생 수 계산
## 1학년부터 6학년까지 더한 값을 총 학생 수로 계산
df['총학생수'] = df.iloc[:, 2:].sum(axis=1)
print(df, '\n')

# 2. 교사 한 명단 맡은 학생 수
## 총 학생 수 / 교사 수로 교사 한 명당 맡은 학생의 수 계산
df['학생/교사'] = df['총학생수'] / df['교사수']
print(df.head(), '\n')

# 3. 학생/교사 최대값의 인덱스명
## 2번의 최댓값을 가진 행을 idxmax() 로 구한다
## idxmax() 활용시 주의점 : '인덱스명'을 반환한다
idx = df['학생/교사'].idxmax()

# 4. 학생/교사 최대값의 학교 교사 수 출력
## 인덱스명으로 값을 찾기 위해서는 loc 활용
## 교사 수를 묻고 있으므로 idxmax() 로 반환받은 인덱스명과 '교사 수' 컬럼을 선택해 값 출력
print(df.loc[idx, '교사수'])

      학교명  교사수  1학년  2학년  3학년  4학년  5학년  6학년  총학생수
0     학교1   20  118   72  154  118   77   59   598
1     학교2   31  187  193  135  137   63   90   805
2     학교3   23  136  193  135  112   91   50   717
3     학교4   32   72  146   67   51  130  164   630
4     학교5   22   61   73   86   82   55  124   481
..    ...  ...  ...  ...  ...  ...  ...  ...   ...
95   학교96   38   75  166  113  191  172   57   774
96   학교97   30  125  169  145  122  187  194   942
97   학교98   37  136  100   93   60  145   63   597
98   학교99   37   74  175  145  157   95   76   722
99  학교100   40   68  109  196  148   84  179   784

[100 rows x 9 columns] 

   학교명  교사수  1학년  2학년  3학년  4학년  5학년  6학년  총학생수      학생/교사
0  학교1   20  118   72  154  118   77   59   598  29.900000
1  학교2   31  187  193  135  137   63   90   805  25.967742
2  학교3   23  136  193  135  112   91   50   717  31.173913
3  학교4   32   72  146   67   51  130  164   630  19.687500
4  학교5   22   61   73   86   82   55  124   481  21.863636 

19


### ✏️ 연도별로 총 범죄 건수(범죄 유형별 발생 건수의 합)의 월평균 값을 구한 후 그 값이 가장 큰 연도를 찾아 해당 연도의 총 범죄 건수의 월평균 값을 출력하시오.(반올림하여 정수 출력)

In [7]:
import pandas as pd
df = pd.read_csv('./data/data6-1-3.csv')
print(df.head())

          날짜  강력범죄  절도범죄  폭력범죄  지능범죄  풍속범죄  교통범죄  경찰서명
0  2020년 01월    22   102    86    62    28   212  B경찰서
1  2020년 02월    26   138    80    61    31   183  E경찰서
2  2020년 03월    14   129    76    60    29   202  C경찰서
3  2020년 04월    26   142    83    71    33   182  B경찰서
4  2020년 05월    28   131    80    72    28   212  B경찰서


In [ ]:
# [풀이1]
# 1. 총 범죄 건수 계산
## 범죄 유형별 발생 건수를 합히고, 새로운 컬럼에 대입 -> 행별 합을 위해 축(axis)을 1로 설정
df['총범죄건수'] = df.iloc[:, 1:7].sum(axis=1)
print(df.head(), '\n')

# 2. 연도 슬라이싱
df['연도'] = df['날짜'].str[:4]
print(df.head(), '\n')

# 3. 연도별 총 범죄 건수 합 계산
## 총 범죄 건수를 연도별로 그룹핑하고 합하기
result = df['총범죄건수'].groupby(df['연도']).sum()
print(result, '\n')

# 4. 가장 큰 값의 월평균 계산
## result 에서 가장 큰 값을 12로 나누고 반올림 후 출력
print(round(result.max()/12))

          날짜  강력범죄  절도범죄  폭력범죄  지능범죄  풍속범죄  교통범죄  경찰서명  총범죄건수
0  2020년 01월    22   102    86    62    28   212  B경찰서    512
1  2020년 02월    26   138    80    61    31   183  E경찰서    519
2  2020년 03월    14   129    76    60    29   202  C경찰서    510
3  2020년 04월    26   142    83    71    33   182  B경찰서    537
4  2020년 05월    28   131    80    72    28   212  B경찰서    551 

          날짜  강력범죄  절도범죄  폭력범죄  지능범죄  풍속범죄  교통범죄  경찰서명  총범죄건수    연도
0  2020년 01월    22   102    86    62    28   212  B경찰서    512  2020
1  2020년 02월    26   138    80    61    31   183  E경찰서    519  2020
2  2020년 03월    14   129    76    60    29   202  C경찰서    510  2020
3  2020년 04월    26   142    83    71    33   182  B경찰서    537  2020
4  2020년 05월    28   131    80    72    28   212  B경찰서    551  2020 

연도
2020    6288
2021    6318
2022    6179
2023    6394
2024    6200
Name: 총범죄건수, dtype: int64 

533


In [ ]:
# [풀이2]
# 1. 총 범죄 건수 계산
## 범죄 유형별 발생 건수를 합히고, 새로운 컬럼에 대입 -> 행별 합을 위해 축(axis)을 1로 설정
df['총범죄건수'] = df.iloc[:, 1:7].sum(axis=1)
print(df.head(), '\n')

# 2. 연도 슬라이싱
df['연도'] = df['날짜'].str[:4]
print(df.head(), '\n')

# 3. 연도별 총 범죄 건수 합 계산 후 내림차순 정렬
result = df['총범죄건수'].groupby(df['연도']).sum()
result = result.sort_values(ascending=False)
print(result, '\n')     # 2023 년도의 총 건수가 가장 많은 것 확인

# 4. 가장 큰 값(첫 번째 행)의 월 평균 계산
print(round(result.iloc[0]/12))

          날짜  강력범죄  절도범죄  폭력범죄  지능범죄  풍속범죄  교통범죄  경찰서명  총범죄건수    연도
0  2020년 01월    22   102    86    62    28   212  B경찰서    512  2020
1  2020년 02월    26   138    80    61    31   183  E경찰서    519  2020
2  2020년 03월    14   129    76    60    29   202  C경찰서    510  2020
3  2020년 04월    26   142    83    71    33   182  B경찰서    537  2020
4  2020년 05월    28   131    80    72    28   212  B경찰서    551  2020 

          날짜  강력범죄  절도범죄  폭력범죄  지능범죄  풍속범죄  교통범죄  경찰서명  총범죄건수    연도
0  2020년 01월    22   102    86    62    28   212  B경찰서    512  2020
1  2020년 02월    26   138    80    61    31   183  E경찰서    519  2020
2  2020년 03월    14   129    76    60    29   202  C경찰서    510  2020
3  2020년 04월    26   142    83    71    33   182  B경찰서    537  2020
4  2020년 05월    28   131    80    72    28   212  B경찰서    551  2020 

연도
2023    6394
2021    6318
2020    6288
2024    6200
2022    6179
Name: 총범죄건수, dtype: int64 

533
